internal/partinfo/partition.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package partinfo
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+)
+
+// Partition contains information for identifying and mounting disk partitions.
+type Partition interface {
+	// Path is the path to the device node file associated to the disk partinfo.
+	Path() string
+	// FSType is the string representing the filesystem type to be used for mounting the partinfo.
+	FSType() string
+	// Label is an optional user-supplied volume name.
+	Label() string
+}
+
+var newPartitionFuncs = []func(*os.File) (Partition, error){
+	newFAT32Partition,
+	newExtPartition,
+}
+
+func newPartition(path string) (Partition, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	for _, newFn := range newPartitionFuncs {
+		if p, err := newFn(f); err != nil {
+			if _, err := f.Seek(0, io.SeekStart); err != nil {
+				return nil, fmt.Errorf("cannot seek: %w", err)
+			}
+			continue
+		} else {
+			return p, nil
+		}
+	}
+
+	return nil, errors.New("unrecognized partition")
```

Comment:
Perhaps "unsupported partition type" ? We know its fine, otherwise the kernel would not show it.